Python/optimizer.c

```diff
@@ -338,6 +339,13 @@ translate_bytecode_to_trace(
         ADD_TO_TRACE(SAVE_IP, (int)(instr - (_Py_CODEUNIT *)code->co_code_adaptive));
         int opcode = instr->op.code;
         uint64_t operand = instr->op.arg;
+        // TODO: EXTENDED_ARG handling
```

Comment:
Will do.

Reply:
This turns out more complicated than I expected. The big question is: **what IP should `SAVE_IP` save?**

To make my current deopt strategy work (re-execute the specialized bytecode, which will fail the same guard and then deopt to the generic bytecode), the saved IP should point to the `EXTENDED_ARG` instruction. But the jump offsets are calculated relative to the jumping opcode. I suppose I could add the number of `EXTENDED_ARG` opcodes to the jump offsets (or subtract, for backward jumps).

But I worry that the error handling code might also need an adjustment -- I don't know if `EXTENDED_ARG` is included in exception handling ranges or if the error handling looks at the last executed bytecode.

I guess for now I can just treat `EXTENDED_ARG` as an unsupported opcode and kick it down the road...

Reply:
Saving the IP of the start of the *instruction* (not codeunit) is the correct thing, I think.
On entering the tier 1 interpreter, it will then get the correct `oparg` as it will execute all the `EXTENDED_ARG`s.

`EXTENDED_ARG` can never cause an error, so it doesn't matter whether they are included in the jump tables, although I think they are anyway.